pallets/parachain-staking/src/delegation_requests.rs

```diff
@@ -311,6 +323,175 @@ impl<T: Config> Pallet<T> {
 		}
 	}
 
+	/// Schedules [DelegationAction::Revoke] for the delegator, towards all delegated collator.
+	/// The last fulfilled request causes the delegator to leave the set of delegators.
+	pub(crate) fn delegator_schedule_revoke_all(
+		delegator: T::AccountId,
+	) -> DispatchResultWithPostInfo {
+		let mut state = <DelegatorState<T>>::get(&delegator).ok_or(<Error<T>>::DelegatorDNE)?;
+		let mut updated_scheduled_requests = vec![];
+		let now = <Round<T>>::get().current;
+		let when = now.saturating_add(T::LeaveDelegatorsDelay::get());
+
+		// it is assumed that a multiple delegations to the same collator does not exist, else this
+		// will cause a bug - the last duplicate delegation update will be the only one applied.
+		let mut existing_revoke_count = 0;
+		for bond in state.delegations.0.clone() {
+			let collator = bond.owner;
+			let mut scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
+			let bonded_amount = state
+				.get_bond_amount(&collator)
+				.ok_or(<Error<T>>::DelegationDNE)?;
+
+			// cancel any existing requests
+			let request =
+				Self::cancel_request_with_state(&delegator, &mut state, &mut scheduled_requests);
+			let request = match request {
+				Some(revoke_req) if matches!(revoke_req.action, DelegationAction::Revoke(_)) => {
+					existing_revoke_count += 1;
+					revoke_req // re-insert the same Revoke request
+				}
+				_ => ScheduledRequest {
+					delegator: delegator.clone(),
+					action: DelegationAction::Revoke(bonded_amount.clone()),
+					when_executable: when,
+				},
+			};
+
+			scheduled_requests.push(request);
+			state.less_total = state.less_total.saturating_add(bonded_amount);
+			updated_scheduled_requests.push((collator, scheduled_requests));
+		}
+
+		if existing_revoke_count == state.delegations.0.len() {
+			return Err(<Error<T>>::DelegatorAlreadyLeaving.into());
+		}
+
+		updated_scheduled_requests
+			.into_iter()
+			.for_each(|(collator, scheduled_requests)| {
+				<DelegationScheduledRequests<T>>::insert(collator, scheduled_requests);
+			});
+
+		<DelegatorState<T>>::insert(delegator.clone(), state);
+		Self::deposit_event(Event::DelegatorExitScheduled {
+			round: now,
+			delegator,
+			scheduled_exit: when,
+		});
+		Ok(().into())
+	}
+
+	/// Cancels every [DelegationAction::Revoke] request for a delegator towards a collator.
+	/// Each delegation must have a [DelegationAction::Revoke] scheduled that must be allowed to be
+	/// executed in the current round, for this function to succeed.
+	pub(crate) fn delegator_cancel_scheduled_revoke_all(
+		delegator: T::AccountId,
+	) -> DispatchResultWithPostInfo {
+		let mut state = <DelegatorState<T>>::get(&delegator).ok_or(<Error<T>>::DelegatorDNE)?;
+		let mut updated_scheduled_requests = vec![];
+
+		// pre-validate that all delegations have a Revoke request.
+		for bond in &state.delegations.0 {
+			let collator = bond.owner.clone();
+			let scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
```

Comment:
This seems problematic to me. It's potentially a very large vec and we look it up for every bond that we have. That makes for a big PoV but also affects CPU time since we iterate over it below.

Is there a bound to the number of bonds a delegator can have? If not, this is arbitrarily large (as written).

We would need a weight hint to (1) properly account for this and (2) provide an upper bound for how many could be handled in one request (this would also imply that it would be impossible to call beyond some threshold).

I think a better design would be to provide a method to revoke a set (vec) of delegations and leave it up to the UI to implement "revoke all".

Reply:
The current upper bound is max `100` delegations, and @crystalin recommended reducing it even further. Iterating over 100 elements wouldn't be CPU intensive at all. The PoV size remains same as before, since the state is read once anyway and the object size is the standard delegator state that we have in the runtime storage.

> I think a better design would be to provide a method to revoke a set (vec) of delegations and leave it up to the UI to implement "revoke all".

That could work but with this PR we are currently trying to support the same functionality as before, just executed differently. Maybe @crystalin has an opinion on whether to move `leave` functionality from runtime to the UI instead.

Reply:
It's not the iteration over the 100 bonds that I'm worried about, it's that we look up `DelegationScheduledRequests` for each one, which I think we established earlier is a max of 350 and then iterate over those.

Even if we are ok with the worst case here, we still need to charge appropriately with our benchmark, which either means a weight hint or we charge for the worst-case.

Reply:
I think weight hint is the solution here. Why is it an issue?

Reply:
It was decided that adding a weight hint would cause too many breaking changes since the extrinsic signature would change for very little benefit.
We then tried to charge a base constant fee, but the `WeightToFee` isn't merged yet, so increasing the weight didn't have an effect.
In the end, we decided to leave it as is.